yt_dlp/extractor/pr0gramm.py

```diff
@@ -110,40 +118,65 @@ def _call_api(self, endpoint, video_id, query={}, note='Downloading API json'):
 
         return data
 
+    @staticmethod
+    def _create_source_url(path):
+        return urljoin('https://img.pr0gramm.com', path)
+
     def _real_extract(self, url):
         video_id = self._match_id(url)
         video_info = traverse_obj(
             self._call_api('get', video_id, {'id': video_id, 'flags': self._maximum_flags}),
             ('items', 0, {dict}))
 
-        source = urljoin('https://img.pr0gramm.com', video_info.get('image'))
+        source = video_info.get('image')
         if not source or not source.endswith('mp4'):
             self.raise_no_formats('Could not extract a video', expected=bool(source), video_id=video_id)
 
         tags = None
         if self._is_logged_in:
-            metadata = self._call_api('info', video_id, {'itemId': video_id})
+            metadata = self._call_api('info', video_id, {'itemId': video_id}, note='Downloading tags')
             tags = traverse_obj(metadata, ('tags', ..., 'tag', {str}))
             # Sorted by "confidence", higher confidence = earlier in list
             confidences = traverse_obj(metadata, ('tags', ..., 'confidence', ({int}, {float})))
             if confidences:
                 tags = [tag for _, tag in sorted(zip(confidences, tags), reverse=True)]
 
+        if video_info.get('variants'):
+            format_traversal = ('variants', ..., {
+                'format_id': ('name', {str}),
+                'url': ('path', {self._create_source_url}),
+                'ext': ('mimeType', {mimetype2ext}),
+                'vcodec': ('codec', {str}),
+                'width': ('width', {int}),
+                'height': ('height', {int}),
+                'bitrate': ('bitRate', {float}),
+                'filesize': ('fileSize', {int}),
+            })
+        else:
+            format_traversal = ((None,), {
+                'url': ('image', {self._create_source_url}),
+                'ext': {lambda _: 'mp4'},
+                'format_id': {lambda _: 'source'},
+                'width': ('width', {int}),
+                'height': ('height', {int}),
+            })
```

Comment:
This's unnecessarily complex imo
```suggestion
        formats = traverse_obj(video_info, ('variants', ..., {
            'format_id': ('name', {str}),
            'url': ('path', {self._create_source_url}),
            'ext': ('mimeType', {mimetype2ext}),
            'vcodec': ('codec', {str}),
            'width': ('width', {int_or_none}),
            'height': ('height', {int_or_none}),
            'bitrate': ('bitRate', {float_or_none}),
            'filesize': ('fileSize', {int_or_none}),
        })) if video_info.get('variants') else [{
            'ext': 'mp4',
            'format_id': 'source',
            **traverse_obj(video_info, {
                'url': ('image', {self._create_source_url}),
                'width': ('width', {int_or_none}),
                'height': ('height', {int_or_none}),
        })]
```

Also, better to use int_or_none/float_or_none. We don't care if the API returns it as string/number as long as we can interpret it.